### ISSUE TITLE
Clarification – CSS reset

### ISSUE THREAD

**Issue by JayPanoz** [edrlab/readium-baseline#2]
What do you mean by CSS reset exactly? I must admit it’s a little bit unclear to me in the Reading System context.

Is it…

1. a CSS reset for styling the Reading System layer only?
2. HTML5’s suggested rendering, kind of fine-tuned for EPUB e.g. pagination and responsiveness? cf. [this stylesheet](https://github.com/FriendsOfEpub/WillThatBeOverriden/blob/master/ReadingSystems/html5/html5.css)
3. something like `normalize.css`, but in reverse, in order to manage some EPUB files’ contents more easily?
4. something like `sanitize.css`, which is going a little bit further than `normalize.css`, in order to “sanitize” problematic CSS in EPUB files? cf. [iBooks’, which is sorta doing that using custom attributes](https://github.com/FriendsOfEpub/WillThatBeOverriden/blob/master/ReadingSystems/iBooks/OS%20X/user_stylesheet_flowable.css.tmpl)
5. a CSS reset for EPUB files’ contents, in order to apply some overrides afterwards? cf. [Overdrive’s Book.css]( https://github.com/FriendsOfEpub/WillThatBeOverriden/blob/master/ReadingSystems/OverDrive/Book.css)

This is an important distinction because some approaches are likely to involve a lot of research + building a test case with various authoring tools’ outputs.

**Comment by HadrienGardeur** [comment]
We can't assume that all EPUB files will have stylesheets and it's always good to have the same kind of CSS reset that you would include in an EPUB for that reason.

Some default styles should also be included for that same reason.

**Comment by JayPanoz** [comment]
OK. 

I guess you won’t be using custom attributes/classes on the fly but rely on the cascade only? (as opposed to [Google Play Books](https://github.com/FriendsOfEpub/WillThatBeOverriden/blob/master/ReadingSystems/Google%20Play%20Books/iOS/gpb.css#L281) for instance, which transforms EPUB files into fragments)

**Comment by HadrienGardeur** [comment]
We also need to define how such styles/stylesheets will be injected (A `<style>` tag or a link? Where do we inject that link?.

But you're right that we absolutely want to avoid mingling with custom attributes/classes if we can, ideally it should be cascade only.

**Comment by JayPanoz** [comment]
Hum OK.

From experience authoring tools tend to export one or several classes e.g. `<p class=".s1 .s17>"`, `<h1 class="chapter-title">`, etc. Sometimes they over qualify selectors in CSS (`h1.chapter-title`) and sometimes they don’t. So that’s `0010` or `0011` in terms of specificity.

And they sometimes leverage `ids` for stuff they want to enforce at any cost e.g. InDesign images’ wrappers (`#_idContainer001`), which is a `0100`.

On the other hand, CSS authors tend to over qualify selectors to get around overrides e.g. `h1 span` or `p.copyright a` (`0002` and `0012`). And, oh well, there’s also `!important`. The worst-case scenario is the use of prefixed properties like `-webkit-text-fill-color` to enforce color of links for instance. As soon as iBooks overrode links’ color in night mode, some people started using it in their CSS—and it can be painful to manage in those modes.

Of course, that’s related to “Handling Overrides” and “Night Mode” but it implies CSS reset + default styles should probably be < `0010`, maybe even < `0003` and `0001` whenever possible (basic typographic styles for unstyled books). 

The biggest issue will obviously be on the essential overrides’ side since images’, svg’s and tables’ sizing, and stuff like `display: inline-block` may depend on and be painful to manage in pagination, and both authoring software and CSS authors don’t usually pay attention to it because other Reading Systems have tackled this issue for them.

I think a list of things which must be overridden so that rendering/display is not screwed up e.g. images or text cut off, etc. could help — it would also probably help for user settings/reading modes handling. As far as I can tell, the CSS reset, and possibly default styles, will depend on every other point in the Todo list. 

As for external (`link`) and internal (`style`), it might be both since external > internal when it comes to specificity.

My biggest concerns at the moment are 

1. properties || values which are not necessarily supported everywhere (e.g. `object-fit`, which solves images’ sizing issue in a simple and elegant manner) and I guess polyfilling is out of question (?)
2. problematic styles like absolute positioning—which some authoring tools are using extensively even if it’s discouraged in specs—, media queries + multicolumns, etc. 
3. all the infrequently-used properties in authors’ CSS hacks which might bite the reader in the a\*\* in some configs (reading modes/settings).

To sum things up, this CSS reset will probably build itself around pagination, handling overrides and night mode. It could at least be a lot easier once those three have been set since you can then design it with specificity levels in mind, which would also tone down the `!importants`.

**Comment by HadrienGardeur** [comment]
What you're describing is exactly what we'd like to avoid with this project: content producers and app developers being at war against one another.

I'd like to have a space where content producers can provide their input and design together a balanced solution for Readium-2.

Purely for pagination, I'm afraid that we won't have a one size fits all solution. So far the two most likely candidates are:

- CSS columns
- overflow: paged-x

You're well aware of the limitations and problems around CSS columns, but for now it's still the solution that works best across a number of platforms.

overflow paged-x provides impressive results performance-wise but has a number of other problems:

- it's not supported on a number of platforms
- there are a bunch of iOS specific bugs
- it's unclear how much control we can have over things like margins, vertical writing and reading direction
- the future of this property is unknown

I wouldn't consider polyfilling as out of question at this point, but if we can avoid it, we definitely should.

**Comment by JayPanoz** [comment]
Yeah please be assured I got that, especially as I wish Reading Systems didn't override to make it up for crappy CSS. I understand your main goal is to prevent Readium 2 to find itself in the comments of  CSS hacks in [such a stylesheet](https://github.com/asciidoctor/asciidoctor-epub3/blob/master/data/styles/epub3.css) and it’s definitely refreshing.

To clarify, when I'm using “essential overrides”, it’s about compensation i.e. normalizing styles which might be altered depending on pagination/scroll implementation so that they behave as authors might expect. 

For instance, sometimes you can find stuff like 

```
<svg … width="100%" height="100%" preserveAspectRatio="xMidYMid meet" … viewBox="0 0 1400 2100"><image… /></svg>
```

Now, authors don’t necessarily specify an height for `html` + `body` because it just works in say RMSDK and iBooks. So what they may get is the SVG’s sizing based on `width="100%"`–and the SVG being cut off. Problem is a lot of authors actually consider it’s a bug on the Reading System’s side while it obviously isn’t since this is how `100% height` should work…

When it comes to `audio`, specifying an `height` (especially `auto`) can have nasty implications in some versions of Webkit i.e. the audio controls won’t display at all. 

The `vh` viewport unit works as intended only because of pagination implementations (e.g. contents loaded in an `iframe`) but it can trigger weird bugs in scroll implementations—personally, I've had issues in Readium (Chrome extension) when using `vh` or `calc` in scroll config e.g. the height grows when additional contents are loaded or a `100vh` cover is `150px` because this is the default for replaced elements.

My goal was just to emphasize that depending on how pagination and scroll are implemented, some kind of “minimalistic normalize” might be designed so that some styles work as expected. Let’s say it’s like providing authors with a bedrock upon which they can build with confidence.

By extension, it could be useful to get some feedback from CSS authors, because they can tell what they expect and what doesn’t currently work as expected. 

As an aside, it feels like a significant part of authors rely on authoring tools so it could also be considered a cultural issue. The toughest part will not necessarily be designing the baseline but documenting and evangelizing it to CSS authors and authoring software developers.

**Comment by HadrienGardeur** [comment]
Native apps won't use an `iframe`, but Web apps won't have any other choice.

I agree that in addition to a normal CSS reset, we'll probably have to apply some minimal normalization too, but this is already trickier.
